bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,976 @@
+	logLabel = "cbt-go-proxy"
+)
+
+// rowToProto converts a Bigtable Go client Row struct into a
+// Bigtable protobuf Row struct. It iterates over all of the column families
+// (keys) and ReadItem slices (values) in the client Row struct
+func rowToProto(btRow bigtable.Row) (*btpb.Row, error) {
+	pbRow := &btpb.Row{
+		Key:      []byte(btRow.Key()),
+		Families: make([]*btpb.Family, 0),
+	}
+
+	for fam, ris := range btRow {
+		pbFam := &btpb.Family{
+			Name:    fam,
+			Columns: make([]*btpb.Column, 0),
+		}
+
+		for _, col := range ris {
+			// Format of column name is `family:columnQualifier`
+			colQualifier := strings.Split(col.Column, ":")[1]
+			pbCol := &btpb.Column{
+				Qualifier: []byte(colQualifier),
+				Cells: []*btpb.Cell{
+					{
+						Value:           col.Value,
+						TimestampMicros: col.Timestamp.Time().UnixMicro(),
+						Labels:          col.Labels,
+					},
+				},
+			}
+			pbFam.Columns = append(pbFam.Columns, pbCol)
+		}
+
+		pbRow.Families = append(pbRow.Families, pbFam)
+	}
+
+	return pbRow, nil
+}
+
+// rowSetFromProto translates a Bigtable v2.RowSet object to a Bigtable.RowSet
+// object.
+func rowSetFromProto(rs *btpb.RowSet) bigtable.RowSet {
+	rowRangeList := make(bigtable.RowRangeList, 0)
+
+	// Convert all rowKeys into single-row RowRanges
+	if rowKeys := rs.GetRowKeys(); len(rowKeys) > 0 {
+		for _, b := range rowKeys {
+
+			// Find the next highest key using byte operations
+			// This allows us to create a RowRange of 1 rowKey
+			e := binary.BigEndian.Uint64(b)
+			e++
+
+			s := binary.Size(e)
+			bOut := make([]byte, s)
+			binary.BigEndian.PutUint64(bOut, e)
+
+			rowRangeList = append(rowRangeList, bigtable.NewRange(string(b), string(bOut)))
+		}
+	}
+
+	if rowRanges := rs.GetRowRanges(); len(rowRanges) > 0 {
+		for _, rrs := range rowRanges {
+			var start, end string
+			var rr bigtable.RowRange
+
+			switch rrs.StartKey.(type) {
+			case *btpb.RowRange_StartKeyClosed:
+				start = string(rrs.GetStartKeyClosed())
+			case *btpb.RowRange_StartKeyOpen:
+				start = string(rrs.GetStartKeyOpen())
+			default:
+				start = ""
+			}
+
+			switch rrs.EndKey.(type) {
+			case *btpb.RowRange_EndKeyClosed:
+				end = string(rrs.GetEndKeyClosed())
+				rr = bigtable.NewRange(start, end)
+			case *btpb.RowRange_EndKeyOpen:
+				end = string(rrs.GetEndKeyOpen())
+				rr = bigtable.NewRange(start, end)
+			default:
+				// If not set, get the infinite row range
+				rr = bigtable.InfiniteRange(start)
+			}
+
+			rowRangeList = append(rowRangeList, rr)
+		}
+	}
+	return rowRangeList
+}
+
+// mutationFromProto translates a slice of Bigtable v2.Mutation objects into
+// a single Bigtable.Mutation object.
+func mutationFromProto(mPbs []*btpb.Mutation) *bigtable.Mutation {
+	m := bigtable.NewMutation()
+	for _, mpb := range mPbs {
+
+		switch mut := mpb.Mutation; mut.(type) {
+		case *btpb.Mutation_DeleteFromColumn_:
+			del := mut.(*btpb.Mutation_DeleteFromColumn_)
```

Comment:
In the context of the case statement, mut will be the type that you matched on. So doing yet another type assertion is unnecessary.

Reply:
Ditto below.

Reply:
I'm afraid the compiler doesn't like it when I remove this type cast, specifically when I attempt to call the methods below.

Reply:
Change the switch statement above to `switch mut := mpb.Mutation.(type) {` and you'll be able to remove these type assertions.